### DISCUSSION TITLE
DJL v0.13.0 release

### DISCUSSION THREAD

**Discussion by frankfliu** [deepjavalibrary/djl#1295] [News]
DJL v0.13.0 brings the new [TensorRT](https://github.com/deepjavalibrary/djl/tree/master/engines/tensorrt) and [Python](https://github.com/deepjavalibrary/djl-serving/tree/master/engines/python) engines, and updates the engines [PyTorch](https://github.com/awslabs/djl/tree/master/engines/pytorch/pytorch-engine) to 1.9.0, [ONNXRuntime](https://github.com/deepjavalibrary/djl/tree/master/engines/onnxruntime/onnxruntime-engine) to 1.9.0, [PaddlePaddle](https://github.com/deepjavalibrary/djl/tree/master/engines/paddlepaddle/paddlepaddle-engine) to 2.0.2, and introduces several new features:

## Key Features

* Introduces [TensorRT](https://github.com/deepjavalibrary/djl/tree/master/engines/tensorrt) engine
* Introduces [Python](https://github.com/deepjavalibrary/djl-serving/tree/master/engines/python) engine which allows you to run Python scripts with DJL
* Upgrades [PyTorch](https://github.com/awslabs/djl/tree/master/engines/pytorch/pytorch-engine) engine to 1.9.0 with CUDA 11.1 support
* Upgrades [ONNXRuntime](https://github.com/deepjavalibrary/djl/tree/master/engines/onnxruntime/onnxruntime-engine) engine to 1.9.0 with UINT8 support
* Upgrades [PaddlePaddle](https://github.com/deepjavalibrary/djl/tree/master/engines/paddlepaddle/paddlepaddle-engine) engine to 2.0.2
* Introduces the [djl-bench](https://snapcraft.io/djlbench) snap package: `sudo snap install djlbench --classic`
* Introduces dynamic batch feature for djl-serving ([#1154](https://github.com/deepjavalibrary/djl/pull/1154))
* DJL serving becomes a standalone repository: https://github.com/deepjavalibrary/djl-serving ([#1170](https://github.com/deepjavalibrary/djl/pull/1170))
* Allows load ModelZoo model using url ([#1120](https://github.com/deepjavalibrary/djl/pull/1120))
* Support .npy and .npz file format ([#1131](https://github.com/deepjavalibrary/djl/pull/1131))
* djl-serving is available in [dockerhub](https://hub.docker.com/r/deepjavalibrary/djl-serving)
* Publishes d2l book Chinese translation preview (chapter 1-5) : https://d2l-zh.djl.ai

## Enhancement

* Introduces several new features in djl-serving:
    * Improves djl-serving API to make it easy to get HTTP headers ([#1134](https://github.com/deepjavalibrary/djl/pull/1134))
    * Loads models on all GPUs at startup for djl-serving ([#1132](https://github.com/deepjavalibrary/djl/pull/1132))
    * Enables asynchronous logging for djl-serving 
    * Makes djl-serving access log in separate log file ([#1150](https://github.com/deepjavalibrary/djl/pull/1150))
    * Adds configuration to support number worker threads for GPU inference ([#1153](https://github.com/deepjavalibrary/djl/pull/1153))
    * Improves auto-scale algorithm for djl-serving ([#1149](https://github.com/deepjavalibrary/djl/pull/1149))
* Introduces several new features in djl-bench:
    * Adds a command line option to djl-bench to generate NDList file ([#1155](https://github.com/deepjavalibrary/djl/pull/1155))
    * Adds warmup to benchmark ([#1152](https://github.com/deepjavalibrary/djl/pull/1152))
    * Improves djll-bench to support djl:// urls ([#1146](https://github.com/deepjavalibrary/djl/pull/1146))
    * Adds support to benchmark on multiple GPUs ([#1144](https://github.com/deepjavalibrary/djl/pull/1144))
    * Adds support to benchmark onnx on GPU machines ([#1148](https://github.com/deepjavalibrary/djl/pull/1148))
    * Adds support to benchmark TensorRT models ([#1257](https://github.com/deepjavalibrary/djl/pull/1257))
    * Adds support to benchmark Python models ([#1267](https://github.com/deepjavalibrary/djl/pull/1267))
* Introduces several new features in PyTorch engine:
    * Supports PyTorch custom input data type with IValue ([#1208](https://github.com/deepjavalibrary/djl/pull/1208))
* Introduces several new features in OnnxRuntime:
    * Adds UINT8 support for OnnxRuntime ([#1271](https://github.com/deepjavalibrary/djl/pull/1271))
* Introduces several new features in PaddlePaddle:
    * Adds more model loading options for PaddlePaddle ([#1173](https://github.com/deepjavalibrary/djl/pull/1173))
    * Adds load functionalities to PaddlePaddle ([#1140](https://github.com/deepjavalibrary/djl/pull/1140))
    * Adds remove pass option to PaddlePaddle ([#1141](https://github.com/deepjavalibrary/djl/pull/1141))
* Introduces several API improvements:
    * Adds missing NDList.get(String) API ([#1194](https://github.com/deepjavalibrary/djl/pull/1194))
    * Adds support to directly load models from a TFHub url ([#1231](https://github.com/deepjavalibrary/djl/pull/1231))
    * Improves repository API to support passing argument in the URL query string ([#1139](https://github.com/deepjavalibrary/djl/pull/1139))
    * Avoids loading the default engine if it is not being used ([#1136](https://github.com/deepjavalibrary/djl/pull/1136))
    * Improves IO by adding a buffer to read/write ([#1135](https://github.com/deepjavalibrary/djl/pull/1135))
    * Improves NDArray.toString() debug mode performance ([#1142](https://github.com/deepjavalibrary/djl/pull/1142))
    * Makes GPU device detection engine specific to avoid confusion when using multiple engines ([#1138](https://github.com/deepjavalibrary/djl/pull/1138))

## Documentation and examples

* Adds Style Transfer example with CycleGAN ([#1180](https://github.com/deepjavalibrary/djl/pull/1180))

## Breaking change

* Removes support for Apache MXNet 1.6.0
* Deprecates Device.getDevices() API - Use Engine.getDevices() instead
* Renames SimpleVocabulary to DefaultVocabulary

## Bug Fixes

* Fixes broken link in documents
* Fixes TensorFlow NDArray was created on CPU instead of GPU bug ([#1279](https://github.com/deepjavalibrary/djl/pull/1279))
* Fixes default image processing pipeline ([#1268](https://github.com/deepjavalibrary/djl/pull/1268))
* Fixed XGBoost NDArray multiple read bug ([#1239](https://github.com/deepjavalibrary/djl/pull/1239))
* Fixes platform matching bug ([#1167](https://github.com/deepjavalibrary/djl/pull/1167))
* Fixes NullPointerException in NDArray.toString() ([#1157](https://github.com/deepjavalibrary/djl/pull/1157))
* Fixes PaddlePaddle crash due to GC ([#1162](https://github.com/deepjavalibrary/djl/pull/1162))
* Fixes NDArrayAdapter.getSparseFormat() unsupported bug ([#1151](https://github.com/deepjavalibrary/djl/pull/1151))
* Fixes mixed device issue in multiple engine use case ([#1123](https://github.com/deepjavalibrary/djl/pull/1123))
* Fixes handle duplicate plugin issue for djl-serving ([#1108](https://github.com/deepjavalibrary/djl/pull/1108))
* Fixes XGBoost NDArray creation bug  ([#1109](https://github.com/deepjavalibrary/djl/pull/1109))
* Fixes runtime exception running benchmark arm machine([#1107](https://github.com/deepjavalibrary/djl/pull/1107))
* Fixes unregister model regression ([#1101](https://github.com/deepjavalibrary/djl/pull/1101))

## Contributors

This release is thanks to the following contributors:

* Akshay Rajvanshi ([@aksrajvanshi](https://github.com/ghost))
* Aziz Zayed ([@AzizZayed](https://github.com/AzizZayed))
* Elchanan Haas ([@ElchananHaas](https://github.com/ElchananHaas))
* Erik Bamberg ([@ebamberg](https://github.com/ebamberg))
* Frank Liu ([@frankfliu](https://github.com/frankfliu))
* Jake Lee ([@stu1130](https://github.com/stu1130))
* Kimi MA ([@kimim](https://github.com/kimim))
* Paul Greyson
* Qing Lan ([@lanking520](https://github.com/lanking520))
* Raymond Liu ([@raymondkhliu](https://github.com/raymondkhliu))
* Sindhu Somasundaram ([@sindhuvahinis](https://github.com/sindhuvahinis))
* Zach Kimberg ([@zachgk](https://github.com/zachgk))

<hr /><em>This discussion was created from the release <a href='https://github.com/deepjavalibrary/djl/releases/tag/v0.13.0'>DJL v0.13.0 release</a>.</em>
